File: src/main/java/com/github/quiltservertools/interdimensional/command/GeneratorCommand.java
package com.github.quiltservertools.interdimensional.command;

import com.github.quiltservertools.interdimensional.command.argument.GeneratorArgumentType;
import com.github.quiltservertools.interdimensional.duck.ServerPlayerEntityAccess;
import com.github.quiltservertools.interdimensional.mixin.ChunkGeneratorAccessor;
import com.mojang.brigadier.arguments.StringArgumentType;
import com.mojang.brigadier.exceptions.CommandSyntaxException;
import com.mojang.brigadier.tree.LiteralCommandNode;
import me.lucko.fabric.api.permissions.v0.Permissions;
import net.minecraft.server.command.ServerCommandSource;
import net.minecraft.util.Identifier;
import net.minecraft.util.registry.Registry;
import net.minecraft.world.biome.source.*;
import net.minecraft.world.dimension.DimensionType;
import net.minecraft.world.gen.chunk.*;
import net.minecraft.world.gen.feature.StructureFeature;

import java.util.*;

import static com.github.quiltservertools.interdimensional.command.InterdimensionalCommand.*;
import static net.minecraft.server.command.CommandManager.argument;
import static net.minecraft.server.command.CommandManager.literal;

public class GeneratorCommand implements Command {
    @Override
    public LiteralCommandNode<ServerCommandSource> register() {
        return literal("generator")
                .requires(Permissions.require("interdimensional.commands.generator", 3))
                .then(argument("args", StringArgumentType.greedyString()).suggests(GeneratorArgumentType.INSTANCE)
                        .executes(ctx -> updateGenerator(ctx.getSource(), GeneratorArgumentType.INSTANCE.rawProperties(StringArgumentType.getString(ctx, "args")))))
                .build();
    }

    private int updateGenerator(ServerCommandSource scs, HashMap<String, Object> propertyMap) throws CommandSyntaxException {
        ChunkGenerator generator = ((ServerPlayerEntityAccess) scs.getPlayer()).getCustomGenerator();
        if (generator == null) generator = scs.getWorld().getChunkManager().getChunkGenerator();
        long seed = scs.getWorld().getSeed();

        if (propertyMap.containsKey("seed")) {
            seed = (long) propertyMap.get("seed");
        }

        // Handle biome sources. Only one is allowed at any one time
        BiomeSource biomeSource = generator.getBiomeSource();
        long biomeSeed = propertyMap.containsKey("biome_seed") ? (long) propertyMap.get("biome_seed") : seed;

        if (propertyMap.containsKey("single_biome")) {
            biomeSource = new FixedBiomeSource(scs.getRegistryManager().get(Registry.BIOME_KEY).get(new Identifier((String) propertyMap.get("single_biome"))));
            scs.sendFeedback(info("Set biome source to Single Biome"), false);

        } else if (propertyMap.containsKey("vanilla_layered")) {
            var largeBiomes = propertyMap.containsKey("large_biomes") && (boolean) propertyMap.get("large_biomes");
            biomeSource = new VanillaLayeredBiomeSource(biomeSeed, false, largeBiomes, scs.getRegistryManager().get(Registry.BIOME_KEY));
            scs.sendFeedback(info("Set biome source to Vanilla Layered"), false);

        } else if (propertyMap.containsKey("multi_noise")) {
            MultiNoiseBiomeSource.method_35242(scs.getRegistryManager().get(Registry.BIOME_KEY), biomeSeed);
            scs.sendFeedback(error("This option is not supported yet"), false);

        } else if (propertyMap.containsKey("the_end_biome_source")) {
            biomeSource = new TheEndBiomeSource(scs.getRegistryManager().get(Registry.BIOME_KEY), biomeSeed);
            scs.sendFeedback(info("Set biome source to End Biome Source"), false);

        } else {
            scs.sendFeedback(info("No biome source option specified, using default"), false);
        }

        ((ChunkGeneratorAccessor) generator).setBiomeSource(biomeSource);

        StructuresConfig structuresConfig;
        // Handle structures
        if (!propertyMap.containsKey("generate_structures") || (boolean) propertyMap.get("generate_structures")) {
            var generateStrongholds = !propertyMap.containsKey("generate_strongholds") || (boolean) propertyMap.get("generate_strongholds");
            if (propertyMap.containsKey("exclude_structures")) {
                structuresConfig = generateStructuresConfig(generateStrongholds, (String) propertyMap.get("exclude_structures"), true);
            } else if (propertyMap.containsKey("include_structures")) {
                structuresConfig = generateStructuresConfig(generateStrongholds, (String) propertyMap.get("include_structures"), false);
            } else {
                structuresConfig = generateStructuresConfig(generateStrongholds, "", true);
            }
            scs.sendFeedback(info("Set structures config to default" + (generateStrongholds ? " with strongholds" : "") + " and overrides"), false);

        } else {
            structuresConfig = new StructuresConfig(Optional.empty(), new HashMap<>());
        }

        ((ChunkGeneratorAccessor) generator).setStructuresConfig(structuresConfig);

        if (!(propertyMap.containsKey("superflat") && (boolean) propertyMap.get("superflat"))) {
            ((ServerPlayerEntityAccess) scs.getPlayer()).setCustomGenerator(generator);
        } else {
            ((ServerPlayerEntityAccess) scs.getPlayer()).setCustomGenerator(new FlatChunkGenerator(new FlatChunkGeneratorConfig(structuresConfig, scs.getRegistryManager().get(Registry.BIOME_KEY))));
        }

        scs.sendFeedback(success("Updated generator configuration"), false);

        return 1;
    }

    private StructuresConfig generateStructuresConfig(boolean generateStrongholds, String list, boolean exclude) {
        Optional<StrongholdConfig> strongholds = generateStrongholds ? Optional.of(StructuresConfig.DEFAULT_STRONGHOLD) : Optional.empty();
        String[] split = list.split(",");
        List<String> strings = new ArrayList<>(Arrays.asList(split));
        Map<StructureFeature<?>, StructureConfig> features = new HashMap<>();
        StructuresConfig.DEFAULT_STRUCTURES.forEach(((structureFeature, structureConfig) -> {
            if ((exclude && !strings.contains(structureFeature.getName())) || (!exclude && strings.contains(structureFeature.getName()))) {
                features.put(structureFeature, structureConfig);
            }
        }));
        return new StructuresConfig(strongholds, features);
    }
}

File: src/main/kotlin/net/quiltservertools/interdimensional/command/GeneratorCommand.kt
package net.quiltservertools.interdimensional.command

import com.mojang.brigadier.arguments.LongArgumentType
import com.mojang.brigadier.context.CommandContext
import com.mojang.brigadier.tree.LiteralCommandNode
import me.lucko.fabric.api.permissions.v0.Permissions
import net.minecraft.command.argument.IdentifierArgumentType
import net.minecraft.command.suggestion.SuggestionProviders
import net.minecraft.server.command.CommandManager.argument
import net.minecraft.server.command.CommandManager.literal
import net.minecraft.server.command.ServerCommandSource
import net.minecraft.server.world.ServerWorld
import net.minecraft.util.registry.Registry
import net.minecraft.world.biome.Biome
import net.minecraft.world.biome.source.FixedBiomeSource
import net.minecraft.world.biome.source.TheEndBiomeSource
import net.minecraft.world.gen.chunk.*
import net.quiltservertools.interdimensional.command.InterdimensionalCommand.error
import net.quiltservertools.interdimensional.command.InterdimensionalCommand.info
import net.quiltservertools.interdimensional.command.argument.ServerDimensionArgument
import net.quiltservertools.interdimensional.customGenerator
import net.quiltservertools.interdimensional.mixin.ChunkGeneratorAccessor

object GeneratorCommand : Command {
    override fun register(): LiteralCommandNode<ServerCommandSource> {
        return literal("generator")
            .requires(Permissions.require("interdimensional.commands.generator", 3))
            .then(literal("clear")
                .executes { resetGenerator(it, null) }
                .then(ServerDimensionArgument.dimension("default")
                    .executes { resetGenerator(it, ServerDimensionArgument.get(it, "default")) }
                )
            )
            .then(
                literal("biome_seed")
                    .then(argument("biome_seed", LongArgumentType.longArg())
                        .executes { updateBiomeSeed(it, LongArgumentType.getLong(it, "biome_seed")) })
            )
            .then(
                literal("biome_source")
                    .then(literal("multi_noise")
                        .then(ServerDimensionArgument.dimension("multi_noise_like")
                            .executes {
                                addMultiNoise(
                                    it,
                                    ServerDimensionArgument.get(it, "multi_noise_like"),
                                    null
                                )
                            }
                            .then(argument("multi_noise_seed", LongArgumentType.longArg())
                                .executes {
                                    addMultiNoise(
                                        it,
                                        ServerDimensionArgument.get(it, "multi_noise_like"),
                                        LongArgumentType.getLong(it, "multi_noise_like")
                                    )
                                }
                            )
                        )
                    )
                    .then(
                        literal("single_biome")
                            .then(
                                argument(
                                    "biome",
                                    IdentifierArgumentType.identifier()
                                ).suggests(SuggestionProviders.ALL_BIOMES)
                                    .executes {
                                        addSingleBiome(
                                            it,
                                            it.source.registryManager.get(Registry.BIOME_KEY)
                                                .get(IdentifierArgumentType.getIdentifier(it, "biome"))
                                        )
                                    }
                            )
                    )
                    .then(
                        literal("end_biome_source")
                            .then(
                                ServerDimensionArgument.dimension("biomes")
                                    .then(argument("seed", LongArgumentType.longArg())
                                        .executes {
                                            addEndBiomeSource(
                                                it,
                                                ServerDimensionArgument.get(it, "biomes"),
                                                LongArgumentType.getLong(it, "seed")
                                            )
                                        }
                                    )
                            )
                            .then(argument("seed", LongArgumentType.longArg())
                                .executes {
                                    addEndBiomeSource(
                                        it,
                                        null,
                                        LongArgumentType.getLong(it, "seed")
                                    )
                                }
                            )
                    )
            )
            .build()
    }

    private fun getGenerator(scs: ServerCommandSource): ChunkGenerator {
        return scs.player.customGenerator ?: scs.world.chunkManager.chunkGenerator
    }

    private fun setGenerator(generator: ChunkGenerator, scs: ServerCommandSource) {
        scs.player.customGenerator = generator
    }

    private fun updateBiomeSeed(ctx: CommandContext<ServerCommandSource>, seed: Long): Int {
        val scs = ctx.source
        val generator = getGenerator(scs)
        val biomeSource = generator.biomeSource.withSeed(seed)
        (generator as ChunkGeneratorAccessor).setBiomeSource(biomeSource)
        setGenerator(generator, scs)
        scs.sendFeedback("Updated biome source to have seed $seed".info(), false)
        return 0
    }

    private fun addMultiNoise(ctx: CommandContext<ServerCommandSource>, noiseLike: ServerWorld, seed: Long?): Int {
        val scs = ctx.source
        var biomeSource = noiseLike.chunkManager.chunkGenerator.biomeSource
        if (seed != null) {
            biomeSource = biomeSource.withSeed(seed)
        }
        val generator = getGenerator(scs)
        (generator as ChunkGeneratorAccessor).setBiomeSource(biomeSource)
        scs.sendFeedback(
            "Set biome source to multi-noise biomes of ${noiseLike.registryKey.value} with seed ${seed ?: noiseLike.seed}".info(),
            false
        )
        return 0
    }

    private fun addSingleBiome(ctx: CommandContext<ServerCommandSource>, biome: Biome?): Int {
        val scs = ctx.source
        if (biome == null) {
            scs.sendFeedback("No biome provided for single biome source".error(), false)
            return 0
        }
        val biomeSource = FixedBiomeSource(biome)
        (getGenerator(scs) as ChunkGeneratorAccessor).setBiomeSource(biomeSource)
        scs.sendFeedback(
            "Set biome source to single biome of type ${
                scs.registryManager.get(Registry.BIOME_KEY).getId(biome)
            }".info(), false
        )
        return 0
    }

    private fun addEndBiomeSource(
        ctx: CommandContext<ServerCommandSource>,
        biomeWorld: ServerWorld?,
        seed: Long?
    ): Int {
        val scs = ctx.source
        val biomes = if (biomeWorld != null) {
            biomeWorld.registryManager.get(Registry.BIOME_KEY)
        } else {
            scs.world.registryManager.get(Registry.BIOME_KEY)
        }
        val generator = getGenerator(scs)

        val biomeSource = TheEndBiomeSource(biomes, seed ?: scs.world.seed)
        (generator as ChunkGeneratorAccessor).setBiomeSource(biomeSource)
        setGenerator(generator, scs)
        return 0
    }

    private fun resetGenerator(ctx: CommandContext<ServerCommandSource>, defaultWorld: ServerWorld?): Int {
        val scs = ctx.source
        scs.player.customGenerator = if (defaultWorld != null) {
            defaultWorld.chunkManager.chunkGenerator
        } else {
            scs.world.chunkManager.chunkGenerator
        }
        return 0
    }
}